### PR TITLE
[#1148] Don't show deleted files in listing

### PR DESCRIPTION
**Pull request by raxod502**
When fd is installed, use it with appropriate options instead of git ls-files in order to fix https://github.com/bbatsov/projectile/issues/1148.

-----------------

Before submitting a PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [x] The commits are consistent with our [contribution guidelines](../blob/master/CONTRIBUTING.md)
- [ ] You've added tests (if possible) to cover your change(s)
- [x] All tests are passing ([`eldev test`](https://github.com/doublep/eldev))
- [x] The new code is not generating bytecode or `M-x checkdoc` warnings
- [x] You've updated the [changelog](../blob/master/CHANGELOG.md) (if adding/changing user-visible functionality)
- [ ] You've updated the readme (if adding/changing user-visible functionality)

Thanks!

### CODE REVIEW

**Comment by raxod502** [comment]
No test or README updates made because this type of functionality was not previously present in test framework or README. However, if it would help out the project, I don't mind making updates for one or both to start them off.

File: projectile.el [hunk 1]
```diff
@@ -1422,7 +1445,12 @@ IGNORED-DIRECTORIES may optionally be provided."
   "Determine which external command to invoke based on the project's VCS.
 Fallback to a generic command when not in a VCS-controlled project."
   (pcase vcs
-    ('git projectile-git-command)
+    ('git (if projectile-git-use-fd
```
Comment (bbatsov) [inline projectile.el:1448]: I think this should also be guarded by `executable-find`, so this won't fail if `fg` is not installed.

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,10 @@
 * Add elm project type.
 * [#1821](https://github.com/bbatsov/projectile/pull/1821): Add `pyproject.toml` discovery for python projects.
 
+### Changes
+
+* [#1285](https://github.com/bbatsov/projectile/pull/1825): By default, use [fd](https://github.com/sharkdp/fd) in Git repositories instead of `git ls-files` when it is installed, in order to solve the problem where deleted files were still shown in `projectile-find-file` until their deletions were staged. The user-facing behavior should be the same, although potentially with different performance characteristics in large Git repositories. The old behavior can be reclaimed by setting `projectile-git-use-fd` to nil.
```
Comment (bbatsov) [inline CHANGELOG.md:12]: I'd also make a note of this in the docs (probably somewhere in usage and in the FAQ).

File: projectile.el [hunk 3]
```diff
@@ -672,11 +672,34 @@ means check all the subdirectories of DIRECTORY.  Etc."
   :type '(repeat (choice directory (cons directory (integer :tag "Depth"))))
   :package-version '(projectile . "1.0.0"))
 
+(defcustom projectile-fd-executable
+  (cond
+   ((executable-find "fdfind") "fdfind")
+   ((executable-find "fd") "fd"))
+  "Path or name of fd executable used by Projectile if enabled.
+Nil means fd is not installed or should not be used."
+  :type 'string)
```
Comment (bbatsov) [inline projectile.el:681]: The new defcustoms should also have a `:package-version` property.

File: projectile.el (hunk 1 above)
Comment (raxod502) [reply projectile.el]: https://github.com/bbatsov/projectile/pull/1825/commits/e49fee486cbb1e3d8caf446c91d057a7ffcb0f93 - guarding by the value of `projectile-fd-executable`, which caches the return value of `executable-find`.

File: CHANGELOG.md (hunk 2 above)
Comment (raxod502) [reply CHANGELOG.md:12]: https://github.com/bbatsov/projectile/pull/1825/commits/e49fee486cbb1e3d8caf446c91d057a7ffcb0f93

File: projectile.el (hunk 3 above)
Comment (raxod502) [reply projectile.el:681]: https://github.com/bbatsov/projectile/pull/1825/commits/e49fee486cbb1e3d8caf446c91d057a7ffcb0f93

File: CHANGELOG.md (hunk 2 above)
Comment (raxod502) [reply CHANGELOG.md:12]: Just realized there is a separate usage page apart from the readme. It already says fd is used when possible and installed, shall I add more detail about how Projectile determines which tool to use?

Comment (bbatsov) [reply CHANGELOG.md:12]: Yeah, I guess that'd be a nice addition to the docs.

File: projectile.el [hunk 4]
```diff
@@ -672,11 +672,35 @@ means check all the subdirectories of DIRECTORY.  Etc."
...
 (defcustom projectile-git-command "git ls-files -zco --exclude-standard"
   "Command used by projectile to get the files in a git project."
   :group 'projectile
   :type 'string)
 
+(defcustom projectile-git-fd-args "-H -0 -E .git -tf --strip-cwd-prefix"
+  "Arguments to fd used to re-implement `git ls-files'.
+This is used with `projectile-fd-executable' when `projectile-git-use-fd'
+is non-nil."
+  :group 'projectile
```
Comment (bbatsov) [inline projectile.el:701]: This one should have a `:package-version` as well.

Comment (raxod502) [reply projectile.el:701]: Sorry, https://github.com/bbatsov/projectile/pull/1825/commits/b68f7efe985f6afb0503f056ae8877dc8a77821b

File: CHANGELOG.md (hunk 2 above)
Comment (raxod502) [reply CHANGELOG.md:12]: https://github.com/bbatsov/projectile/pull/1825/commits/b68f7efe985f6afb0503f056ae8877dc8a77821b

**Comment by raxod502** [comment]
Looks like CI is failing with:

```
/home/runner/work/_actions/purcell/setup-emacs/master/dist/install-nix.sh: line 63: cachix: command not found
```

at e.g. https://github.com/bbatsov/projectile/actions/runs/4299885959/jobs/7495540856

I don't think that's related to this change. I'm happy to help troubleshoot that if it would be helpful.

**Comment by raxod502** [comment]
Looks like the above error was resolved, however snapshot build is failing with:

```
             projectile.el:6134:19: Error: reference to free variable ‘savehist-additional-variables’
[00:00.346]  projectile.el:6134:19: Error: assignment to free variable ‘savehist-additional-variables’
```

Again I don't think it's related to this change, but happy to help troubleshoot.

**Comment by bbatsov** [comment]
Thanks for working on this. I'll check the CI problem myself at some point.

**Pull request merged**